File: LeetCode_Marathons/LCM6/Hard/Patching_Array/a.cpp
class Solution {
public:
    int minPatches(vector<int>& nums, int n) {
        long long dpSum = 0;
        int counter = 0;
        nums.push_back(n);
        
        for (int num: nums) {
            while (dpSum + 1 < num) {
                if (dpSum >= n) return counter;
                counter++;
                dpSum = dpSum + 1 + dpSum;
            }
            dpSum += num;
            if (dpSum >= n) return counter;
        }
        return counter;
    }
};

File: LeetCode_Marathons/LCM6/Hard/Patching_Array/a.go
func minPatches(nums []int, n int) int {
    dpSum := 0
    ans := 0
    nums = append(nums, n + 1)
    for _, num := range nums {
        for dpSum + 1 < num {
            if dpSum >= n {
                return ans
            }
            ans++
            dpSum = dpSum * 2 + 1
        }
        dpSum += num
        if dpSum >= n {
            return ans
        }
    }
    return ans
}